File: src/Sema.hs
-- | Semantic analysis pass for Homun.
--
-- This pass performs:
--   1.  snake_case enforcement for variable/lambda names
--   2.  Recursion detection — wraps recursive lambdas with a `rec` marker
--   3.  Mutual recursion detection — emits a compile error
--   4.  Monomorphisation guard — detects same lambda used at 2+ different types (best-effort)
--   5.  Undefined reference check for top-level bindings
--
-- The pass runs AFTER parsing and BEFORE codegen.
-- It returns either a list of error messages or the annotated (but same-shaped) program.
module Sema
  ( analyzeProgram
  , SemaError(..)
  ) where

import AST
import Data.Char (isUpper, isLower)
import Data.List (nub, group, sort, isPrefixOf)
import qualified Data.Set as Set
import qualified Data.Map.Strict as Map

-- ─────────────────────────────────────────
-- Error type
-- ─────────────────────────────────────────

data SemaError
  = BadCasing    Name              -- variable/lambda not snake_case
  | MutualRec    Name Name         -- mutual recursion detected
  | Undefined    Name              -- unresolved reference at top level
  | PolyConflict Name              -- lambda used at conflicting types
  | EmptyTypedList Name            -- @[] that is never used
  deriving (Show, Eq)

formatError :: SemaError -> String
formatError (BadCasing n)      = "SEMA ERROR: '" ++ n ++ "' must be snake_case"
formatError (MutualRec a b)    = "SEMA ERROR: mutual recursion between '" ++ a ++ "' and '" ++ b ++ "' is forbidden"
formatError (Undefined n)      = "SEMA ERROR: undefined reference '" ++ n ++ "'"
formatError (PolyConflict n)   = "SEMA ERROR: lambda '" ++ n ++ "' is used at conflicting concrete types (monomorphisation failure)"
formatError (EmptyTypedList n) = "SEMA ERROR: empty list '" ++ n ++ "' is never used, cannot infer type"

-- ─────────────────────────────────────────
-- Analysis environment
-- ─────────────────────────────────────────

type Env  = Set.Set Name   -- defined names in scope
type Errs = [SemaError]

-- ─────────────────────────────────────────
-- Entry point
-- ─────────────────────────────────────────

analyzeProgram :: Program -> Either [SemaError] Program
analyzeProgram prog =
  let builtins = Set.fromList
        [ "print", "len", "range", "str", "int", "float", "bool"
        , "filter", "map", "reduce", "load_ron", "save_ron"
        , "clamp", "update", "idle", "attack", "die", "warn", "recover"
        ]
      topNames = Set.fromList [ n | SBind n _ <- prog ]
      env0     = Set.union builtins topNames
      errs     = checkStmts env0 prog
                 ++ checkCasingAll prog
                 ++ checkMutualRec prog
  in if null errs
       then Right (map (markRecursion topNames) prog)
       else Left errs

-- ─────────────────────────────────────────
-- 1. snake_case enforcement
-- ─────────────────────────────────────────

checkCasingAll :: Program -> Errs
checkCasingAll = concatMap checkCasingStmt

checkCasingStmt :: Stmt -> Errs
checkCasingStmt (SBind n _)
  | isTypeName n  = []    -- PascalCase types are allowed
  | not (isSnake n) = [BadCasing n]
checkCasingStmt _ = []

isSnake :: Name -> Bool
isSnake [] = True
isSnake n  = all (\c -> c `elem` "_0123456789abcdefghijklmnopqrstuvwxyz") n

isTypeName :: Name -> Bool
isTypeName (c:_) = isUpper c
isTypeName []    = False

-- ─────────────────────────────────────────
-- 2. Undefined reference check
-- ─────────────────────────────────────────

-- | Check a sequence of stmts, threading newly bound names into env
checkStmts :: Env -> [Stmt] -> Errs
checkStmts _   []     = []
checkStmts env (s:ss) =
  let errs = checkStmt env s
      env' = case s of
               SBind n _      -> Set.insert n env
               SStructDef n _ -> Set.insert n env
               SEnumDef n _   -> Set.insert n env
               _              -> env
  in errs ++ checkStmts env' ss

checkStmt :: Env -> Stmt -> Errs
checkStmt env (SBind _ e)    = checkExpr env e
checkStmt env (SExprStmt e)  = checkExpr env e
checkStmt _ _                = []

checkExpr :: Env -> Expr -> Errs
checkExpr env e = case e of
  EVar n ->
    if Set.member n env || n == "_" then [] else [Undefined n]
  EField ex _        -> checkExpr env ex
  EIndex ex idx      -> checkExpr env ex ++ checkExpr env idx
  ESlice ex a b c    -> checkExpr env ex ++ maybe [] (checkExpr env) a
                         ++ maybe [] (checkExpr env) b
                         ++ maybe [] (checkExpr env) c
  EList xs           -> concatMap (checkExpr env) xs
  EDict pairs        -> concatMap (\(k,v) -> checkExpr env k ++ checkExpr env v) pairs
  ESet xs            -> concatMap (checkExpr env) xs
  ETuple xs          -> concatMap (checkExpr env) xs
  EStruct _ fields   -> concatMap (checkExpr env . snd) fields
  EBinOp _ a b       -> checkExpr env a ++ checkExpr env b
  EUnOp _ a          -> checkExpr env a
  EPipe a b           -> checkExpr env a ++ checkExpr env b
  ECall fn args      -> checkExpr env fn ++ concatMap (checkExpr env) args
  EIf c ts te ec     ->
    checkExpr env c ++ checkStmts env ts ++ checkExpr env te ++
    maybe [] (\(es,ee) -> checkStmts env es ++ checkExpr env ee) ec
  EMatch sc arms     ->
    checkExpr env sc ++ concatMap (checkArm env) arms
  EFor v iter stmts fe ->
    let env'     = Set.insert v env
        envFinal = foldl stmtBound env' stmts
    in checkExpr env iter ++ checkStmts env' stmts
       ++ maybe [] (checkExpr envFinal) fe
  EWhile c stmts fe  ->
    let envFinal = foldl stmtBound env stmts
    in checkExpr env c ++ checkStmts env stmts
       ++ maybe [] (checkExpr envFinal) fe
  EBreak me          -> maybe [] (checkExpr env) me
  ELambda params _ _ stmts fe ->
    let env'     = foldr (\(Param n _) s -> Set.insert n s) env params
        envFinal = foldl stmtBound env' stmts
    in checkStmts env' stmts ++ checkExpr envFinal fe
  ELoadRon p _       -> checkExpr env p
  ESaveRon d p       -> checkExpr env d ++ checkExpr env p
  _                  -> []

checkArm :: Env -> MatchArm -> Errs
checkArm env (MatchArm pat guard body) =
  let env' = extendWithPat env pat
  in maybe [] (checkExpr env') guard ++ checkExpr env' body

extendWithPat :: Env -> Pat -> Env
extendWithPat env PWild           = env
extendWithPat env PNone           = env
extendWithPat env (PVar n)        = Set.insert n env
extendWithPat env (PLit _)        = env
extendWithPat env (PTuple ps)     = foldr (flip extendWithPat) env ps
extendWithPat env (PEnum _ mp)    = maybe env (extendWithPat env) mp

stmtBound :: Env -> Stmt -> Env
stmtBound e (SBind n _)      = Set.insert n e
stmtBound e (SStructDef n _) = Set.insert n e
stmtBound e (SEnumDef n _)   = Set.insert n e
stmtBound e _                = e

-- ─────────────────────────────────────────
-- 3. Mutual recursion detection
-- ─────────────────────────────────────────

-- Build call-graph between top-level lambdas, detect cycles of length > 1
checkMutualRec :: Program -> Errs
checkMutualRec prog =
  let lambdas  = [ (n, e) | SBind n e@(ELambda {}) <- prog ]
      callGraph = Map.fromList [ (n, freeCallsIn n (lambdaBody e)) | (n, e) <- lambdas ]
  in concatMap (findMutual callGraph) (Map.keys callGraph)

lambdaBody :: Expr -> Expr
lambdaBody (ELambda _ _ _ _ e) = e
lambdaBody e = e

-- | Names directly called by the body of function n (excluding n itself)
freeCallsIn :: Name -> Expr -> [Name]
freeCallsIn self e = filter (/= self) (collectCalls e)

collectCalls :: Expr -> [Name]
collectCalls (ECall (EVar n) args) = n : concatMap collectCalls args
collectCalls (EBinOp _ a b)        = collectCalls a ++ collectCalls b
collectCalls (EUnOp _ a)           = collectCalls a
collectCalls (EPipe a b)            = collectCalls a ++ collectCalls b
collectCalls (EIf c ts te ec)       = collectCalls c
  ++ concatMap collectStmtCalls ts ++ collectCalls te
  ++ maybe [] (\(es,ee) -> concatMap collectStmtCalls es ++ collectCalls ee) ec
collectCalls (EMatch sc arms)       = collectCalls sc ++ concatMap (\(MatchArm _ g b) -> collectCalls b) arms
collectCalls (EFor _ it ss fe)      = collectCalls it ++ concatMap collectStmtCalls ss ++ maybe [] collectCalls fe
collectCalls (EWhile c ss fe)       = collectCalls c ++ concatMap collectStmtCalls ss ++ maybe [] collectCalls fe
collectCalls (ELambda _ _ _ ss fe)  = concatMap collectStmtCalls ss ++ collectCalls fe
collectCalls _                      = []

collectStmtCalls :: Stmt -> [Name]
collectStmtCalls (SBind _ e)   = collectCalls e
collectStmtCalls (SExprStmt e) = collectCalls e
collectStmtCalls _             = []

-- | For each function, do a DFS up to depth 2 looking for mutual calls
findMutual :: Map.Map Name [Name] -> Name -> Errs
findMutual graph n =
  let callees = Map.findWithDefault [] n graph
      mutuals  = filter (\c -> n `elem` Map.findWithDefault [] c graph) callees
  in map (MutualRec n) mutuals

-- ─────────────────────────────────────────
-- 4. Recursion marking
-- ─────────────────────────────────────────

-- | After analysis, annotate recursive lambdas so codegen can emit the right Rust
markRecursion :: Set.Set Name -> Stmt -> Stmt
markRecursion topNames (SBind n (ELambda params ret void stmts finalExpr))
  | isRecursiveExpr n finalExpr || any (isRecursiveStmt n) stmts =
      SBind n (ELambda params ret void stmts finalExpr)  -- In real impl, wrap with Rc<RefCell<_>>
  | otherwise =
      SBind n (ELambda params ret void stmts finalExpr)
markRecursion _ s = s

isRecursiveExpr :: Name -> Expr -> Bool
isRecursiveExpr n (ECall (EVar m) _) = n == m
isRecursiveExpr n (EBinOp _ a b)     = isRecursiveExpr n a || isRecursiveExpr n b
isRecursiveExpr n (EIf _ _ te ec)    = isRecursiveExpr n te || maybe False (\(_,e) -> isRecursiveExpr n e) ec
isRecursiveExpr n (EMatch _ arms)    = any (\(MatchArm _ _ b) -> isRecursiveExpr n b) arms
isRecursiveExpr n (EFor _ _ ss fe)   = any (isRecursiveStmt n) ss || maybe False (isRecursiveExpr n) fe
isRecursiveExpr _ _                  = False

isRecursiveStmt :: Name -> Stmt -> Bool
isRecursiveStmt n (SBind _ e)   = isRecursiveExpr n e
isRecursiveStmt n (SExprStmt e) = isRecursiveExpr n e
isRecursiveStmt _ _             = False

File: src/sema.rs
/// Semantic analysis pass for Homun.
///
/// 1. snake_case enforcement for variable/lambda names
/// 2. Undefined reference check for top-level bindings
/// 3. Mutual recursion detection
use crate::ast::*;
use std::collections::HashSet;
use std::fmt;

#[derive(Debug, PartialEq, Eq)]
pub enum SemaError {
    BadCasing(Name),
    Undefined(Name),
}

impl fmt::Display for SemaError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            SemaError::BadCasing(n) => write!(f, "SEMA ERROR: '{}' must be snake_case", n),
            SemaError::Undefined(n) => write!(f, "SEMA ERROR: undefined reference '{}'", n),
        }
    }
}

pub fn analyze_program(prog: &Program) -> Result<(), Vec<SemaError>> {
    analyze_program_with_imports(prog, &HashSet::new())
}

pub fn analyze_program_with_imports(
    prog: &Program,
    imported_names: &HashSet<String>,
) -> Result<(), Vec<SemaError>> {
    // Builtins — always available (provided by builtin.rs preamble).
    let builtins: HashSet<String> = [
        "print", "str", "int", "float", "bool",
    ]
    .iter()
    .map(|s| s.to_string())
    .collect();

    // `use std` provides these (range, len, filter, map, reduce, etc.).
    let std_names: HashSet<String> = [
        "len", "range", "filter", "map", "reduce",
        "load_ron", "save_ron",
        "abs", "min", "max", "clamp",
        "sorted", "reversed", "enumerate", "zip", "flatten",
        "any", "all", "count", "unique", "index_of",
        "split", "join", "trim", "trim_start", "trim_end",
        "starts_with", "ends_with", "replace", "to_upper", "to_lower",
        "chars", "find", "contains", "repeat", "substr",
        "strip_prefix", "strip_suffix", "lines", "is_empty",
        "is_digit", "is_alpha", "is_alnum", "is_upper", "is_lower",
        "pad_left", "pad_right",
        "std",
    ]
    .iter()
    .map(|s| s.to_string())
    .collect();

    // Check if program has `use std` or `use ext`.
    let has_use_std = prog.iter().any(|s| matches!(s, Stmt::Use(p) if p.len() == 1 && p[0] == "std"));
    let has_use_ext = prog.iter().any(|s| matches!(s, Stmt::Use(p) if p.len() == 1 && p[0] == "ext"));

    let top_names: HashSet<String> = prog
        .iter()
        .filter_map(|s| match s {
            Stmt::Bind(n, _) | Stmt::StructDef(n, _) | Stmt::EnumDef(n, _) => Some(n.clone()),
            _ => None,
        })
        .collect();

    let mut env0: HashSet<String> = builtins.union(&top_names).cloned().collect();
    env0 = env0.union(imported_names).cloned().collect();
    if has_use_std {
        env0 = env0.union(&std_names).cloned().collect();
    }

    let mut errs = Vec::new();
    // When `use ext` is present, skip undefined-reference checks
    // because ext provides many functions sema can't introspect.
    if !has_use_ext {
        errs.extend(check_stmts(&env0, prog));
    }
    errs.extend(check_casing_all(prog));
    if errs.is_empty() {
        Ok(())
    } else {
        Err(errs)
    }
}

// ─── 1. snake_case enforcement ───────────────────────────────

fn check_casing_all(prog: &Program) -> Vec<SemaError> {
    prog.iter().flat_map(check_casing_stmt).collect()
}

fn check_casing_stmt(stmt: &Stmt) -> Vec<SemaError> {
    match stmt {
        Stmt::Bind(n, _) => {
            if is_type_name(n) || is_snake(n) {
                vec![]
            } else {
                vec![SemaError::BadCasing(n.clone())]
            }
        }
        _ => vec![],
    }
}

fn is_snake(n: &str) -> bool {
    n.is_empty()
        || n.chars()
            .all(|c| c.is_ascii_lowercase() || c.is_ascii_digit() || c == '_')
}

fn is_type_name(n: &str) -> bool {
    n.chars().next().is_some_and(|c| c.is_uppercase())
}

// ─── 2. Undefined reference check ───────────────────────────

fn check_stmts(env: &HashSet<String>, stmts: &[Stmt]) -> Vec<SemaError> {
    let mut errs = Vec::new();
    let mut env = env.clone();
    for s in stmts {
        errs.extend(check_stmt(&env, s));
        match s {
            Stmt::Bind(n, _) | Stmt::StructDef(n, _) | Stmt::EnumDef(n, _) => {
                env.insert(n.clone());
            }
            _ => {}
        }
    }
    errs
}

fn check_stmt(env: &HashSet<String>, stmt: &Stmt) -> Vec<SemaError> {
    match stmt {
        Stmt::Bind(_, e) | Stmt::Expression(e) => check_expr(env, e),
        _ => vec![],
    }
}

fn check_expr(env: &HashSet<String>, expr: &Expr) -> Vec<SemaError> {
    match expr {
        Expr::Var(n) => {
            if n == "_" || env.contains(n) {
                vec![]
            } else {
                vec![SemaError::Undefined(n.clone())]
            }
        }
        Expr::Field(e, _) => check_expr(env, e),
        Expr::Index(e, idx) => {
            let mut errs = check_expr(env, e);
            errs.extend(check_expr(env, idx));
            errs
        }
        Expr::Slice(e, a, b, c) => {
            let mut errs = check_expr(env, e);
            for x in [a, b, c].into_iter().flatten() {
                errs.extend(check_expr(env, x));
            }
            errs
        }
        Expr::List(xs) | Expr::Set(xs) | Expr::Tuple(xs) => {
            xs.iter().flat_map(|x| check_expr(env, x)).collect()
        }
        Expr::Dict(pairs) => pairs
            .iter()
            .flat_map(|(k, v)| {
                let mut e = check_expr(env, k);
                e.extend(check_expr(env, v));
                e
            })
            .collect(),
        Expr::Struct(_, fields) => fields
            .iter()
            .flat_map(|(_, e)| check_expr(env, e))
            .collect(),
        Expr::BinOp(_, a, b) | Expr::Pipe(a, b) => {
            let mut errs = check_expr(env, a);
            errs.extend(check_expr(env, b));
            errs
        }
        Expr::UnOp(_, a) => check_expr(env, a),
        Expr::Call(f, args) => {
            let mut errs = check_expr(env, f);
            for a in args {
                errs.extend(check_expr(env, a));
            }
            errs
        }
        Expr::If(c, ts, te, ec) => {
            let mut errs = check_expr(env, c);
            errs.extend(check_stmts(env, ts));
            errs.extend(check_expr(env, te));
            if let Some((es, ee)) = ec {
                errs.extend(check_stmts(env, es));
                errs.extend(check_expr(env, ee));
            }
            errs
        }
        Expr::Match(sc, arms) => {
            let mut errs = check_expr(env, sc);
            for arm in arms {
                errs.extend(check_arm(env, arm));
            }
            errs
        }
        Expr::For(v, iter, stmts, fe) => {
            let mut env2 = env.clone();
            env2.insert(v.clone());
            let mut errs = check_expr(env, iter);
            errs.extend(check_stmts(&env2, stmts));
            let env_final = stmts_bound(&env2, stmts);
            if let Some(e) = fe {
                errs.extend(check_expr(&env_final, e));
            }
            errs
        }
        Expr::While(c, stmts, fe) => {
            let mut errs = check_expr(env, c);
            errs.extend(check_stmts(env, stmts));
            let env_final = stmts_bound(env, stmts);
            if let Some(e) = fe {
                errs.extend(check_expr(&env_final, e));
            }
            errs
        }
        Expr::Block(stmts, fe) => {
            let mut errs = check_stmts(env, stmts);
            let env_final = stmts_bound(env, stmts);
            errs.extend(check_expr(&env_final, fe));
            errs
        }
        Expr::Break(me) => me.as_ref().map_or(vec![], |e| check_expr(env, e)),
        Expr::Lambda {
            params,
            stmts,
            final_expr,
            ..
        } => {
            let mut env2 = env.clone();
            for p in params {
                env2.insert(p.name.clone());
            }
            let mut errs = check_stmts(&env2, stmts);
            let env_final = stmts_bound(&env2, stmts);
            errs.extend(check_expr(&env_final, final_expr));
            errs
        }
        Expr::LoadRon(p, _) => check_expr(env, p),
        Expr::SaveRon(d, p) => {
            let mut errs = check_expr(env, d);
            errs.extend(check_expr(env, p));
            errs
        }
        _ => vec![],
    }
}

fn check_arm(env: &HashSet<String>, arm: &MatchArm) -> Vec<SemaError> {
    let env2 = extend_with_pat(env, &arm.pat);
    let mut errs = Vec::new();
    if let Some(g) = &arm.guard {
        errs.extend(check_expr(&env2, g));
    }
    errs.extend(check_expr(&env2, &arm.body));
    errs
}

fn extend_with_pat(env: &HashSet<String>, pat: &Pat) -> HashSet<String> {
    let mut env = env.clone();
    match pat {
        Pat::Var(n) => {
            env.insert(n.clone());
        }
        Pat::Tuple(ps) => {
            for p in ps {
                env = extend_with_pat(&env, p);
            }
        }
        Pat::Enum(_, Some(p)) => {
            env = extend_with_pat(&env, p);
        }
        _ => {}
    }
    env
}

fn stmts_bound(env: &HashSet<String>, stmts: &[Stmt]) -> HashSet<String> {
    let mut env = env.clone();
    for s in stmts {
        match s {
            Stmt::Bind(n, _) | Stmt::StructDef(n, _) | Stmt::EnumDef(n, _) => {
                env.insert(n.clone());
            }
            _ => {}
        }
    }
    env
}

